Mathlib/Algebra/Field/Defs.lean

```diff
@@ -47,93 +45,180 @@ a `GroupWithZero` lemma instead.
+
+`↑q : K` is defined as `(q.num : K) * (q.den : K)⁻¹`.
+
+Do not use this directly (instances of `DivisionSemiring` are allowed to override that default for
+better definitional properties). Instead, use the coercion. -/
+def NNRat.castRec [NatCast K] [Div K] (q : ℚ≥0) : K := q.num / (q.den : K)
+
+/-- The default definition of the scalar multiplication by `ℚ≥0` on a division semiring `K`.
+
+`q • x` is defined as `↑q * x`.
+
+Do not use directly (instances of `DivisionSemiring` are allowed to override that default for
+better definitional properties). Instead use the `•` notation. -/
+def nnqsmulRec (coe : ℚ≥0 → K) [Mul K] (a : ℚ≥0) (x : K) : K := coe a * x
+
+/-- The default definition of the coercion `ℚ → K` for a division ring `K`.
+
+`↑q : K` is defined as `(q.num : K) * (q.den : K)⁻¹`.
+
+Do not use this directly (instances of `DivisionRing` are allowed to override that default for
+better definitional properties). Instead, use the coercion. -/
 def Rat.castRec [NatCast K] [IntCast K] [Mul K] [Inv K] : ℚ → K
   | ⟨a, b, _, _⟩ => ↑a * (↑b)⁻¹
 #align rat.cast_rec Rat.castRec
 
-/-- The default definition of the scalar multiplication `(a : ℚ) • (x : K)` for a division ring `K`
-is given by `a • x = (↑ a) * x`.
-Use `(a : ℚ) • (x : K)` instead of `qsmulRec` for better definitional behaviour.
--/
+/-- The default definition of the scalar multiplication by `ℚ` on a division ring `K`.
+
+`q • x` is defined as `↑q * x`.
+
+Do not use directly (instances of `DivisionRing` are allowed to override that default for
+better definitional properties). Instead use the `•` notation. -/
 def qsmulRec (coe : ℚ → K) [Mul K] (a : ℚ) (x : K) : K :=
   coe a * x
 #align qsmul_rec qsmulRec
 
-/-- A `DivisionSemiring` is a `Semiring` with multiplicative inverses for nonzero elements. -/
-class DivisionSemiring (α : Type*) extends Semiring α, GroupWithZero α
+/-- A `DivisionSemiring` is a `Semiring` with multiplicative inverses for nonzero elements.
+
+An instance of `DivisionSemiring K` includes maps `nnratCast : ℚ≥0 → K` and `nnqsmul : ℚ≥0 → K → K`.
+Those two fields are needed to implement the `DivisionSemiring K → Algebra ℚ≥0 K` instance since we
+need to control the specific definitions for some special cases of `K` (in particular `K = ℚ≥0`
+itself). See also note [forgetful inheritance].
+
+If the division semiring has positive characteristic `p`, our division by zero convention forces
+`nnratCast (1 / p) = 1 / 0 = 0`. -/
+class DivisionSemiring (α : Type*) extends Semiring α, GroupWithZero α, NNRatCast α where
+  protected nnratCast := NNRat.castRec
+  /-- However `NNRat.cast` is defined, it must be equal to `a / b`.
+
+  Do not use this lemma directly. Use `NNRat.cast_def` instead. -/
+  protected nnratCast_def (q : ℚ≥0) : (NNRat.cast q : α) = q.num / q.den := by intros; rfl
+  /-- Scalar multiplication by a nonnegative rational number.
+
+  Do not use directly. Instead use the `•` notation. -/
+  protected nnqsmul : ℚ≥0 → α → α := nnqsmulRec NNRat.cast
```

Comment:
If #11262 goes ahead, this should change to
```suggestion
  protected nnqsmul : ℚ≥0 → α → α
```
along with a comment advising to use `nnqsmulRec _`.